### PR TITLE
Allow multiple reads / writes in a stream

### PR DESCRIPTION
**Pull request by lionel-faber**

### CODE REVIEW

File: src/connections.rs [hunk 1]
```diff
@@ -43,76 +39,71 @@ impl Connection {
...
+
+    /// Send message to peer and await for a reponse.
+    pub async fn send(&self, msg: Bytes) -> Result<ConnectionStreams> {
+        let mut connection_streams = self.create_connection_streams().await?;
+        connection_streams.send(msg).await?;
+        Ok(connection_streams)
     }
 
     /// Send message to peer using a bi-directional stream without awaiting for a reponse.
     pub async fn send_only(&self, msg: Bytes) -> Result<()> {
```
Comment (bochaco) [inline src/connections.rs:62]: I'm wondering if we should just remove this function now, if the user wants to only send on a bi-stream then `send` can be used and the result returned can be simply dropped by the user obtaining the same result?

File: src/connections.rs [hunk 2]
```diff
@@ -43,76 +39,71 @@ impl Connection {
         self.quic_conn.remote_address()
     }
 
-    /// Send message to peer and await for a reponse.
-    pub async fn send(&mut self, msg: Bytes) -> Result<Bytes> {
+    /// Get connection streams for reading/writing
+    pub async fn create_connection_streams(&self) -> Result<ConnectionStreams> {
```
Comment (bochaco) [inline src/connections.rs:43]: I was thinking we may want to return two separate objects here so the user can spawn tasks to send and receive on the bi-stream from separate tasks/threads.

Comment (lionel1704) [reply src/connections.rs:43]: Yes that makes sense :+1:

File: src/connections.rs (hunk 1 above)
Comment (joshuef) [reply src/connections.rs:62]: That'd still give you unnecessary wait time, no?

For `send` to work, you'd have to `await` the response, regardless of if you're going to use it?

Comment (lionel1704) [reply src/connections.rs:62]: That's changed now. send will return the streams so you can do `recv.read()` if you want the response or just drop them if you don't expect a response.

File: src/connections.rs [hunk 3]
```diff
@@ -240,38 +191,70 @@ impl IncomingMessages {
...
-                }
-            }
-            Err(err) => {
-                warn!("Failed reading message's bytes: {}", err);
-                None
-            }
+/// Read the message's bytes which size is capped
+pub async fn read_bytes(recv: &mut quinn::RecvStream, _max_msg_size: usize) -> Result<Bytes> {
+    let mut data_len: [u8; 1] = [0; 1];
+    recv.read_exact(&mut data_len).await?;
```
Comment (Yoga07) [inline src/connections.rs:205]: Why do we call `read_exact` twice(here with a slice and below with a vec)? Recon it's for the flag?

File: src/connections.rs [hunk 4]
```diff
@@ -240,38 +191,70 @@ impl IncomingMessages {
...
+        WireMsg::UserMsg(msg_bytes) => Ok(Bytes::copy_from_slice(&msg_bytes)),
+        WireMsg::EndpointEchoReq | WireMsg::EndpointEchoResp(_) => {
+            // TODO: handle the echo request/response message
+            unimplemented!("echo message type not supported yet");
         }
     }
 }
 
+/// Helper to send bytes to peer using the provided stream.
+pub async fn send_msg(
```
Comment (Yoga07) [inline src/connections.rs:219]: Should this be `pub`?

**Review by Yoga07 — COMMENTED** [review]
Looks neat to me! Just minor ones for clarification.

File: src/connections.rs (hunk 4 above)
Comment (lionel1704) [reply src/connections.rs]: I made it public temporarily. In the latest commits I've made it private again.

File: src/connections.rs (hunk 3 above)
Comment (bochaco) [reply src/connections.rs:186]: This is to obtain the length of the message. 

Although, @lionel1704, I think this is where we should use the `WireMsg`, that struct is already being used for prefixing the message bytes with information which is handled only by the quic-p2p layer, same as the length we need now, so we should add a `msg_length` field (fixed size) to `WireMsg` and do a first read from the stream of length == size_of(<`WireMsg` struct>). Once you read that, you parse it with `WireMsg::from_raw(data)` and from there you obtain the info about message type and message length. Then if what you have is `WireMsg::UserMsg({ msg_bytes, msh_length })` you can read a second time from the stream with a slice/Vec of size == msg_length.

**Review by bochaco — CHANGES_REQUESTED** [review]

File: src/connections.rs (hunk 3 above)
Comment (lionel1704) [reply src/connections.rs:186]: >  do a first read from the stream of length == size_of(<WireMsg struct>)

But, this length isn't fixed. So we wouldn't really know how many bytes to read?

Comment (bochaco) [reply src/connections.rs:186]: You are right, the msg bytes should be removed from it, so the WireStruct should have only the length field.

**Review by joshuef — APPROVED** [review]

**Pull request merged**